File: app/src/main/java/com/martinlaizg/geofind/views/fragment/single/TourFragment.java
package com.martinlaizg.geofind.views.fragment.single;

import android.app.AlertDialog;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.Navigation;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.chip.Chip;
import com.martinlaizg.geofind.R;
import com.martinlaizg.geofind.config.Preferences;
import com.martinlaizg.geofind.data.access.api.error.ErrorType;
import com.martinlaizg.geofind.data.access.database.entities.Place;
import com.martinlaizg.geofind.data.access.database.entities.Tour;
import com.martinlaizg.geofind.data.access.database.entities.User;
import com.martinlaizg.geofind.data.enums.PlayLevel;
import com.martinlaizg.geofind.views.adapter.PlaceListAdapter;
import com.martinlaizg.geofind.views.fragment.creator.CreatorFragment;
import com.martinlaizg.geofind.views.fragment.play.PlayMapFragment;
import com.martinlaizg.geofind.views.viewmodel.TourViewModel;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class TourFragment
		extends Fragment {

	public static final String TOUR_ID = "TOUR_ID";

	@BindView(R.id.tour_name)
	TextView tour_name;
	@BindView(R.id.tour_description)
	TextView tour_description;
	@BindView(R.id.tour_creator)
	Chip tour_creator;
	@BindView(R.id.tour_num_places)
	TextView tour_num_places;
	@BindView(R.id.edit_button)
	MaterialButton edit_button;
	@BindView(R.id.places_list)
	RecyclerView places_list;
	@BindView(R.id.other_places)
	RecyclerView other_places;

	@BindView(R.id.completed_text)
	TextView completed_text;
	@BindView(R.id.completed_divider)
	View completed_divider;

	@BindView(R.id.in_progress_text)
	TextView in_progress_text;
	@BindView(R.id.in_progress_divider)
	View in_progress_divider;

	@BindView(R.id.play_button)
	MaterialButton play_button;

	@BindView(R.id.empty_text)
	TextView empty_text;

	private PlaceListAdapter adapterCompleted;
	private PlaceListAdapter adapterNoCompleted;
	private AlertDialog alert;
	private TourViewModel viewModel;
	private User user;

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
			Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_tour, container, false);
		ButterKnife.bind(this, view);
		return view;
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(requireActivity());
		user = Preferences.getLoggedUser(sp);

		adapterCompleted = new PlaceListAdapter(true);
		places_list.setLayoutManager(new LinearLayoutManager(requireActivity()));
		places_list.setAdapter(adapterCompleted);

		adapterNoCompleted = new PlaceListAdapter(false);
		other_places.setLayoutManager(new LinearLayoutManager(requireActivity()));
		other_places.setAdapter(adapterNoCompleted);

		int tour_id = 0;
		Bundle b = getArguments();
		if(b != null) {
			tour_id = b.getInt(TOUR_ID);
		}
		if(tour_id == 0) {
			Toast.makeText(requireContext(),
			               requireContext().getString(R.string.tour_not_permitted),
			               Toast.LENGTH_SHORT).show();
			Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
					.popBackStack();
		}
		viewModel = ViewModelProviders.of(requireActivity()).get(TourViewModel.class);
		viewModel.getTour(tour_id, user.getId()).observe(this, this::setTour);
	}

	private void setTour(Tour tour) {
		if(tour != null) {
			tour_name.setText(tour.getName());
			tour_description.setText(tour.getDescription());
			tour_creator.setText(tour.getCreator().getUsername());
			if(tour.getPlaces().isEmpty()) {
				empty_text.setVisibility(View.VISIBLE);
			}
			if(user != null && user.getId() == tour.getCreator_id()) {
				Bundle b = new Bundle();
				b.putInt(CreatorFragment.TOUR_ID, tour.getId());
				edit_button.setOnClickListener(
						Navigation.createNavigateOnClickListener(R.id.toEditCreator, b));
				edit_button.setVisibility(View.VISIBLE);
			}

			setPlaces(tour);
		} else {
			ErrorType error = viewModel.getError();
			if(error == null) {
				Toast.makeText(requireContext(), getString(R.string.something_went_wrong),
				               Toast.LENGTH_SHORT).show();
				Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
						.popBackStack();
			}
		}
	}

	private void setPlaces(Tour tour) {
		List<Place> places = new ArrayList<>(tour.getPlaces());
		List<Place> playPlaces = viewModel.getPlayPlaces();
		if(playPlaces.size() == 0) {
			// Not played yet
			in_progress_text.setText(getString(R.string.places));
			completed_divider.setVisibility(View.GONE);
			completed_text.setVisibility(View.GONE);
			places_list.setVisibility(View.GONE);
			return;
		}

		// In progress
		int numTotalPlaces = places.size();
		tour_num_places.setText(getResources()
				                        .getQuantityString(R.plurals.number_place, numTotalPlaces,
				                                           numTotalPlaces));
		adapterCompleted.setPlaces(playPlaces);

		for(int i = 0; i < places.size(); i++) {
			for(Place p : playPlaces) {
				if(places.get(i).getId() == p.getId()) {
					places.remove(i);
					i--;
					break;
				}
			}
		}

		adapterNoCompleted.setPlaces(places);
		if(places.size() == 0) {
			play_button.setText(getString(R.string.completed));
			play_button.setEnabled(false);
			in_progress_text.setVisibility(View.GONE);
			in_progress_divider.setVisibility(View.GONE);
		}

		play_button.setOnClickListener(v -> alert.show());
		setDifficultyDialog(tour.getId(), tour.getMin_level());
	}

	private void setDifficultyDialog(int tour_id, PlayLevel min_level) {
		Bundle playArguments = new Bundle();
		playArguments.putInt(PlayMapFragment.TOUR_ID, tour_id);

		// Get all of the difficulties
		List<String> difficulties = new ArrayList<>(
				Arrays.asList(getResources().getStringArray(R.array.difficulties)));
		int numDiff = difficulties.size();
		// Get only the allowed difficulties
		difficulties = difficulties.subList(min_level.ordinal(), difficulties.size());

		// Copy difficulties
		String[] items = new String[difficulties.size()];
		for(int i = 0; i < difficulties.size(); i++) {
			items[i] = difficulties.get(i);
		}

		int offset = numDiff - items.length;
		AlertDialog.Builder builder = new AlertDialog.Builder(requireContext());
		builder.setTitle(getResources().getString(R.string.select_play_difficulty));
		builder.setItems(items, (dialog, item) -> {
			int[] destinations = {R.id.toPlayTour, R.id.toPlayCompass, R.id.toPlayTherm};
			int destPos = item + offset;
			Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
					.navigate(destinations[destPos], playArguments);
		});
		alert = builder.create();
	}
}

File: app/src/main/java/com/martinlaizg/geofind/views/fragment/single/TourFragment.kt
package com.martinlaizg.geofind.views.fragment.single

import android.app.AlertDialog
import android.content.DialogInterface
import android.os.Bundle
import android.text.method.ScrollingMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.Navigation
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.button.MaterialButton
import com.google.android.material.chip.Chip
import com.martinlaizg.geofind.R
import com.martinlaizg.geofind.config.Preferences
import com.martinlaizg.geofind.data.access.database.entities.Place
import com.martinlaizg.geofind.data.access.database.entities.Tour
import com.martinlaizg.geofind.data.access.database.entities.User
import com.martinlaizg.geofind.data.enums.PlayLevel
import com.martinlaizg.geofind.views.adapter.PlaceListAdapter
import com.martinlaizg.geofind.views.fragment.creator.CreatorFragment
import com.martinlaizg.geofind.views.fragment.play.PlayTourFragment
import com.martinlaizg.geofind.views.viewmodel.TourViewModel
import java.util.*

class TourFragment : Fragment() {

	private var tourName: TextView? = null
	private var tourDescription: TextView? = null
	private var tourCreator: Chip? = null
	private var tourNumPlaces: TextView? = null
	private var editButton: MaterialButton? = null
	private var placesList: RecyclerView? = null
	private var otherPlaces: RecyclerView? = null
	private var completedText: TextView? = null
	private var completedDivider: View? = null
	private var inProgressText: TextView? = null
	private var inProgressDivider: View? = null
	private var playButton: MaterialButton? = null
	private var emptyText: TextView? = null

	private var adapterCompleted: PlaceListAdapter? = null
	private var adapterNoCompleted: PlaceListAdapter? = null
	private var alert: AlertDialog? = null
	private var viewModel: TourViewModel? = null
	private var user: User? = null
	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
	                          savedInstanceState: Bundle?): View? {
		val view = inflater.inflate(R.layout.fragment_tour, container, false)
		tourName = view.findViewById(R.id.tour_name)
		tourDescription = view.findViewById(R.id.tour_description)
		tourCreator = view.findViewById(R.id.tour_creator)
		tourNumPlaces = view.findViewById(R.id.tour_num_places)
		editButton = view.findViewById(R.id.edit_button)
		placesList = view.findViewById(R.id.places_list)
		otherPlaces = view.findViewById(R.id.other_places)
		completedText = view.findViewById(R.id.completed_text)
		completedDivider = view.findViewById(R.id.completed_divider)
		inProgressText = view.findViewById(R.id.in_progress_text)
		inProgressDivider = view.findViewById(R.id.in_progress_divider)
		playButton = view.findViewById(R.id.play_button)
		emptyText = view.findViewById(R.id.empty_text)
		tourDescription!!.movementMethod = ScrollingMovementMethod()
		return view
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		val sp = PreferenceManager.getDefaultSharedPreferences(requireActivity())
		user = Preferences.getLoggedUser(sp)
		adapterCompleted = PlaceListAdapter(true)
		placesList!!.layoutManager = LinearLayoutManager(requireActivity())
		placesList!!.adapter = adapterCompleted
		adapterNoCompleted = PlaceListAdapter(false)
		otherPlaces!!.layoutManager = LinearLayoutManager(requireActivity())
		otherPlaces!!.adapter = adapterNoCompleted
		var tourId = 0
		val b = arguments
		if (b != null) {
			tourId = b.getInt(TOUR_ID)
		}
		if (tourId == 0) {
			Toast.makeText(requireContext(),
					requireContext().getString(R.string.tour_not_permitted),
					Toast.LENGTH_SHORT).show()
			Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
					.popBackStack()
		}
		viewModel = ViewModelProviders.of(requireActivity()).get(TourViewModel::class.java)
		viewModel!!.getTour(tourId, user!!.id).observe(viewLifecycleOwner, Observer { tour: Tour? -> setTour(tour) })
	}

	private fun setTour(tour: Tour?) {
		if (tour != null) {
			tourName!!.text = tour.name
			tourDescription!!.text = tour.description
			tourCreator!!.text = tour.creator!!.username
			if (tour.places.isEmpty()) {
				emptyText!!.visibility = View.VISIBLE
			}
			if (user != null && user!!.id == tour.creatorId) {
				val b = Bundle()
				b.putInt(CreatorFragment.TOUR_ID, tour.id)
				editButton!!.setOnClickListener(
						Navigation.createNavigateOnClickListener(R.id.toEditCreator, b))
				editButton!!.visibility = View.VISIBLE
			}
			setPlaces(tour)
		} else {
			val error = viewModel!!.error
			if (error == null) {
				Toast.makeText(requireContext(), getString(R.string.something_went_wrong), Toast.LENGTH_SHORT).show()
				Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
						.popBackStack()
			}
		}
	}

	private fun setPlaces(tour: Tour) {
		val totalPlaces: List<Place> = ArrayList(tour.places)
		val playedPlaces = viewModel!!.playPlaces
		val numTotalPlaces = totalPlaces.size
		tourNumPlaces!!.text = resources
				.getQuantityString(R.plurals.number_place, numTotalPlaces,
						numTotalPlaces)
		playButton!!.setOnClickListener { alert!!.show() }
		setDifficultyDialog(tour.id, tour.minLevel)
		if (playedPlaces.isEmpty()) {
			// Not played yet
			inProgressText!!.text = getString(R.string.places)
			adapterNoCompleted!!.places = totalPlaces
			completedDivider!!.visibility = View.GONE
			completedText!!.visibility = View.GONE
			placesList!!.visibility = View.GONE
			return
		}

		// In progress
		adapterCompleted!!.places = playedPlaces
		val playedPlacesIds = playedPlaces.map(Place::id)
		val noCompletedPlaces = totalPlaces.filter { p -> playedPlacesIds.contains(p.id) }
		adapterNoCompleted!!.places = noCompletedPlaces
		if (totalPlaces.isEmpty()) {
			playButton!!.text = getString(R.string.completed)
			playButton!!.isEnabled = false
			inProgressText!!.visibility = View.GONE
			inProgressDivider!!.visibility = View.GONE
		}
	}

	private fun setDifficultyDialog(tourId: Int, minLevel: PlayLevel?) {
		val playArguments = Bundle()
		playArguments.putInt(PlayTourFragment.TOUR_ID, tourId)

		// Get all of the difficulties
		var difficulties: List<String> = ArrayList(
				listOf(*resources.getStringArray(R.array.difficulties)))
		val numDiff = difficulties.size
		// Get only the allowed difficulties
		difficulties = difficulties.subList(minLevel!!.ordinal, difficulties.size)

		// Copy difficulties
		val items = arrayOfNulls<String>(difficulties.size)
		for (i in difficulties.indices) {
			items[i] = difficulties[i]
		}
		val offset = numDiff - items.size
		val builder = AlertDialog.Builder(requireContext())
		builder.setTitle(resources.getString(R.string.select_play_difficulty))
		builder.setItems(items) { _: DialogInterface?, item: Int ->
			val destinations = intArrayOf(R.id.toPlayTour, R.id.toPlayCompass, R.id.toPlayTherm)
			val destPos = item + offset
			Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
					.navigate(destinations[destPos], playArguments)
		}
		alert = builder.create()
	}

	companion object {
		const val TOUR_ID = "TOUR_ID"
	}
}